src/main/scala/com/fulcrumgenomics/commons/collection/ParIterator.scala

```diff
@@ -0,0 +1,147 @@
+  }
+
+  /**
+    * Constructs a ParIterator from the underlying iterator, that will perform operations in parallel on chunks
+    * of `chunkSize` using the provided `TaskSupport`.
+    *
+    * @param iter the underlying iterator to parallelize over
+    * @param chunkSize the number of elements to chunk together for parallel processing
+    * @param support the TaskSupport to use for parallel computations like `map()`
+    * @param chunkBuffer if Some(n) use an [[AsyncIterator]] to pre-buffer n chunks from the input
+    *                    read for parallel computation
+    */
+  def apply[A](iter: Iterator[A], chunkSize: Int, support: TaskSupport, chunkBuffer: Option[Int]): ParIterator[A] = {
+    val grouped = {
+      val g = iter.grouped(chunkSize)
+      chunkBuffer match {
+        case None    => g
+        case Some(n) => AsyncIterator(g, Some(n))
+      }
+    }
+
+    new ParIterator[A](grouped, chunkSize, support)
+  }
+}
+
+
+/**
+  * A parallel iterator that operates by grouping or chunking the underlying iterator and then performing
+  * operations on the chunks in parallel.  Optionally supports wrapping the iterator in an [[AsyncIterator]]
+  * to draw items through the parallel processing and make them available for consumption while the next chunk(s)
+  * are processed.
+  *
+  * For example:
+  *
+  * {{{
+  * val iter = Io.readLines(reallyBigFile)
+  * ParIterator(iter, chunkSize=4096, threads=8)
+  *   .filter(_.contains("foo"))
+  *   .map(_.toUpperCase)
+  *   .map(_.reverse)
+  *   .toAsync(cache=4096 * 8)
+  *   .foreach(println)
+  * }}}
+  */
+class ParIterator[A] private (private val iter: Iterator[Seq[A]],
+                              val chunkSize: Int = 1024,
+                              private val taskSupport: TaskSupport) extends AbstractIterator[A] {
+  private var currentChunk: Seq[A] = if (iter.hasNext) iter.next() else Seq.empty
+
+  /** True if there are more items available, false otherwise. */
+  override def hasNext: Boolean = currentChunk.nonEmpty || iter.hasNext
+
+  /** Retrieve the next item.  Will throw an exception if there are no more items. */
+  override def next(): A = {
+    while (currentChunk.isEmpty && iter.hasNext) currentChunk = iter.next()
+    val v = currentChunk.head
+    currentChunk = currentChunk.tail
+    v
+  }
+
+  /** Returns an iterator over all remaining chunks including the current chunk. */
+  private def chunks: Iterator[Seq[A]] = Iterator(currentChunk) ++ iter
+
+  /** Wraps an async iterator around this parallel iterator to draw items through the iterator. This should only
+    * be invoked after and other transforming operations such as map/filter/etc.
+    *
+    * @param cache how many elements to accumulate in the async iterator's cache - for best memory usage this
+    *              should be a multiple of chunkSize.
+    */
+  def toAsync(cache: Int = 4 * chunkSize): AsyncIterator[A] = AsyncIterator(this, Some(cache))
```

Comment:
part of me wants the async iterator to be able to use the threads in the pool.  Not sure how feasible that is given AysncIterator is a daemon thread/runnable

Reply:
I think that would be really hard.